### PR TITLE
Vision Transformer

### PR DESCRIPTION
**Pull request by lemairecarl**

### CODE REVIEW

File: article/_posts/2020-10-15-vision_transformer.md [hunk 1]
```diff
@@ -0,0 +1,80 @@
+---
+layout: review
+title: "An Image is Worth 16x16 Words: Transformers for Image Recognition at Scale"
+tags: cnn transformer classification
+author: "Carl Lemaire"
+cite:
+    authors: "Anonymous (Google?)"
```
Comment (jhlegarreta) [inline article/_posts/2020-10-15-vision_transformer.md:7]: I would not imagine about the authors. Even more so if this is a double blind review. If approved, we can submit a separate PR to honor the authors.

File: article/_posts/2020-10-15-vision_transformer.md [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+pdf: "https://openreview.net/forum?id=YicbFdNTTy"
+---
+
+
+# Highlights
+
+* The transformer architecture, known for its performance on NLP tasks, is applied to image classification
+* The proposed architecture, "ViT" (Vision Transformer) is shown to perform as well or better than CNNs for image classification on large scale datasets
+* The usefulness/superiority of the proposed transformer over CNNs only appears when the number of images in the dataset reaches about 100 million
+* The proposed architecture uses about 2X less compute to train, compared to CNNs, for a given performance
```
Comment (jhlegarreta) [inline article/_posts/2020-10-15-vision_transformer.md:19]: 2X less time? 2X less memory?

And related to the above statement, does the resource efficiency statement assume that the training set size is the same for the CNN vs ViT?

File: article/_posts/2020-10-15-vision_transformer.md [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+
+* ImageNet
+* ImageNet ReaL ("Reassessed Labels", from Beyer et al. 2020) [Code](https://github.com/google-research/reassessed-imagenet)
+* CIFAR10/100
+* Oxford Pets, Oxford Flowers
+* VTAB (Zhai et al., 2019b) ("VTAB evaluates low-data transfer using 1 000 examples to diverse tasks.  The tasks are divided into three groups: Natural– tasks like the above, Pets, CIFAR, etc. Specialized– medical and satellite imagery, and Structured– tasks that require geometric understanding like localization.") [Blog post](https://ai.googleblog.com/2019/11/the-visual-task-adaptation-benchmark.html)
+
+# Results
+
+A seen in the table below, ViT performs slightly better to a very large ResNet, and does so using significantly less (fine-tuning) compute.
```
Comment (jhlegarreta) [inline article/_posts/2020-10-15-vision_transformer.md:54]: As seen

Comment (jhlegarreta) [inline article/_posts/2020-10-15-vision_transformer.md:54]: slightly better than ?

Comment (jhlegarreta) [inline article/_posts/2020-10-15-vision_transformer.md:54]: using significantly less (fine-tuning) computational resources?

Not sure what the fine-tuning explanation means in here/how it relates to the computational resources.

File: article/_posts/2020-10-15-vision_transformer.md [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+
+![](/article/images/vit/tab2.jpg)
+
+However, the pre-training has to involve a very large number of training samples: when this number exceeds 100 million, ViT starts to shine. Else, the ResNet performs better. See below:
+
+![](/article/images/vit/fig3-4.jpg)
+
+ViT also compares favourably in terms of pre-training compute, as seen below:
```
Comment (jhlegarreta) [inline article/_posts/2020-10-15-vision_transformer.md:62]: in terms of pre-training computational resources?

File: article/_posts/2020-10-15-vision_transformer.md [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+
+The "Hybrid" approach uses CNN feature vectors as tokens; it is not considered very important by the authors.
+
+The figures below serve to inspect the vision transformer architecture:
+
+![](/article/images/vit/fig7.jpg)
+
+# Conclusions
+
+The Vision Transformer is an architecture that can outperform CNNs given datasets in the 100M-image range. It required less compute to train than the CNNs used in this paper.
```
Comment (jhlegarreta) [inline article/_posts/2020-10-15-vision_transformer.md:74]: less computational resources?

File: article/_posts/2020-10-15-vision_transformer.md [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+
+# Remarks
+
+* [This video](https://www.youtube.com/watch?v=TrdevFK_am4) is a good commentary on the paper.
+    - It shows how double-blind is broken, and that we can easily infer that this paper comes from Google, and which authors it's most likely from
```
Comment (jhlegarreta) [inline article/_posts/2020-10-15-vision_transformer.md:79]: Same comment as above concerning hypothesizing about the authors.

File: article/_posts/2020-10-15-vision_transformer.md (hunk 2 above)
Comment (AntoineTheb) [inline article/_posts/2020-10-15-vision_transformer.md:19]: 2X less ? Like half ?

File: article/_posts/2020-10-15-vision_transformer.md [hunk 7]
```diff
@@ -0,0 +1,80 @@
+---
+layout: review
+title: "An Image is Worth 16x16 Words: Transformers for Image Recognition at Scale"
+tags: cnn transformer classification
```
Comment (AntoineTheb) [inline article/_posts/2020-10-15-vision_transformer.md:4]: `cnn` gives an error, the proper tag is `CNN`

File: article/_posts/2020-10-15-vision_transformer.md [hunk 8]
```diff
@@ -0,0 +1,80 @@
...
+* Images are split into patches (16x16 yields the best results)
+* The patches are flattened, and become the tokens (or "words")
+* The flattened patches are projected using a MLP
+* Each flattened patch is concatenated with a vector which represents the position of the patch. A positional embedding is learned, which maps a one-dimensional patch index to a vector representation.
+* Note that this allows the very first layer to attend to any part of the image, in comparison with CNNs for which the receptive field develops over many layers.
+* The first token is a dummy, and is only there because the feature vector of the image will appear at that position at the last layer of the network.
+
+## Experiment design
+
+* The models are pre-trained on either one of ImageNet (1k classes, 1.3M images), ImageNet-21k (21k classes, 14M images) or JFT (18k classes, 303M images).
```
Comment (AntoineTheb) [inline article/_posts/2020-10-15-vision_transformer.md:37]: pre-trained on either ~one of~ ImageNet

**Review by AntoineTheb — CHANGES_REQUESTED** [review]

**Review by jodoin — APPROVED** [review]

File: article/_posts/2020-10-15-vision_transformer.md (hunk 2 above)
Comment (lemairecarl) [reply article/_posts/2020-10-15-vision_transformer.md:19]: I've seen "compute" many times to refer to the quantity of computations. But I will replace with FLOPS, to make sure there is not confusion.

For your second question, I am not sure. I think yes. But the most important point to me is that the validation set and validation accuracy are the same for all models.

File: article/_posts/2020-10-15-vision_transformer.md (hunk 3 above)
Comment (lemairecarl) [reply article/_posts/2020-10-15-vision_transformer.md:54]: There are two phases: pre-training and fine-tuning. Both phases requires different amounts of FLOPS
